module/zfs/dbuf.c

```diff
@@ -5240,6 +5281,8 @@ dbuf_remap(dnode_t *dn, dmu_buf_impl_t *db, dmu_tx_t *tx)
 	if (!spa_feature_is_active(spa, SPA_FEATURE_DEVICE_REMOVAL))
 		return;
 
+	mutex_enter(&db->db_mtx);
+	rw_enter(&db->db_rwlock, RW_READER);
```

Comment:
This is called only for meta-dnode or indirects, so we should not need the `db_mtx`.  About `db_rwlock` I am not sure. I think by this time modifications of this block should be completed, and old locking inside `dbuf_remap_impl()` should cover our modifications from other readers.